### ISSUE TITLE
Example Usage

### ISSUE THREAD

**Issue by mdhaber** [MrShoenel/metrics-as-scores#5]
One of the items in the JOSS review criteria #1 is:

> Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).

The software appears to have been developed primarily to analyze "Qualitas.class". I don't believe that satisfies this criterion because:
- This analysis appears to be too complex to describe in a step-by-step form
- The criterion suggests that there should be "examples" (plural), and it appears that the software has only been tested on the "Qualitas.class" dataset. 

Can you develop a simpler example usage tutorial? 

IIUC the key terms correctly, perhaps the following scenario is applicable. Note that words (e.g. metric) are used here in the colloquial sense; it is not necessarily correct to interpret these words according to the definitions used in the context of this software.

> I teach a senior design class in which several teams of students designed and built small cranes for lifting and moving heavy loads (e.g. 2000 lbf to a height of 10 ft and within ~10ft radius). We wanted to compare the achievement of different cranes based on their performance in a series of standardized tests: maximum load lifted, maximum height, ease of use, etc. After the cranes completed the tests, we ranked the teams' performance in each test separately. For instance, suppose three teams competed in the "maximum load lifted" test, and the max loads they lifted were:
> - Team 1: 1600 lb
> - Team 2: 2100 lb
> - Team 3: 1850 lb
> We ranked the team's performance in each test (e.g. in the maximum load test, Team 2 ranked 1st, Team 3 2nd, and Team 1 3rd). After ranking all tests in this way, I observed that one team (say Team 2) ranked 1st in four tests, 2nd in one test, and 3rd in one test. Subjectively, I determined that Team 2 was the overall winner based on these rankings.
>
> Such a procedure is probably fine for such a small, informal competition (in which there were only 6 teams and 6 tests). However, I imagined that techniques from statistics might be used to improve the fairness of my competition. For instance, suppose that teams from every mechanical engineering department in the country were to design cranes and perform the same set of tests. In this case, I would have a fairly good estimate of the distribution of values that might be achieved a maximum load test (comparing against a relevant population). The next year, if teams from my department were to design cranes and compete in the same test, I would be able to use the percentile of the maximum load lifted by the crane as a score. Effectively, I'd be ranking them against a large population of relevant teams rather than merely ranking them w.r.t. the five other teams in my department. The scores in all six test would now be on a standardized scale (0 - 1), so perhaps then I could come up with an objective criterion to aggregate the scores from multiple tests in order to choose an overall winner of the competition.

1. What relationships, if any, can you see between your software and this scenario? (If none, perhaps we need a different scenario for the next two questions.)
2. Can you help me interpret the capabilities of this software in terms of this scenario or a similarly simple, concrete scenario? 
3. Can you add documentation demonstrating the usage of key features of this software in such a context (even with synthetic raw data)?

Thanks!

**Comment by MrShoenel** [comment]
I will include usage examples tomorrow.
Metrics As Scores has changed considerably now, and I invite you to try the new CLI (TUI) and test a public dataset or to create your own dataset, perhaps based on the example you posted previously.

After installing the package (or the project using `poetry`), you can bring up the CLI by typing `mas`.

However, I will update the readme tomorrow and include a simple usage scenario there, too.

**Comment by MrShoenel** [comment]
(Example) usage scenarios have been included in the `README.md`. Also, there is now a 3rd new dataset (`diamonds`).
I have tried to make clear why we call it **scores**.
We do not intend to change the name of the application, so I clarified why (and how) we used the terms (software) metrics and scores.

As for your example:

> For instance, suppose that teams from every mechanical engineering department in the country were to design cranes and perform the same set of tests. In this case, I would have a fairly good estimate of the distribution of values that might be achieved a maximum load test (comparing against a relevant population). The next year, if teams from my department were to design cranes and compete in the same test, I would be able to use the percentile of the maximum load lifted by the crane as a score.

That is a scenario that Metrics As Scores would be able to provide insights. In the README, I wrote that Metrics As Scores can be thought of as a "multiple ANOVA". Let me add the following aspect to your example:

- Instead of just the lifted weight, you are likely interested in additional properties of the cranes (e.g., weight, platform size, etc.)

In your case, the departments are the **groups**, and the cranes' properties are **features**. Suppose you want to compare a typical crane from department A to one in department B. The approach that we suggest for that is to estimate (cumulative) densities for each attribute in each group. While a CDF only tells us what the probability of finding a value less than or equal to `x` is, in some cases, an attribute has a meaning (e.g., for two cranes that can lift the same weight, the one which weighs less is preferred). With this, you would perhaps find out that the score you deduct for a crane in department A is much higher than it is in department B because their cranes tend to be a lot heavier while giving similar performance.

The weight in this example has a utopian (most desirable) value of `0`: A weight-less crane, although practically impossible, would be perfect. In other words, the lower the weight, the better.
However, not all attributes have such an **ideal** value. One core concept of Metrics As Scores was to use non-parametric statistical values, such as the expectation, median, etc. of a random variable, and then to transform the observed distribution into a distribution of distances from that ideal value. In the weight example, we could also say that the ideal value is the infimum of all cranes observed to date.
However, suppose we use the expectation instead. Then we could use the CCDF of the transformed distribution to rank how far off future cranes are. This would allow us to identify extremely heavy or light cranes.

About the second aspect: Suppose that we want to compare the groups (here: department). We could ask now, how statistically significantly different groups are when we consider all the available features. For example, suppose that next to the weight, we have five additional attributes. When computing all pair-wise combinations, we may find that it is only the weight that varies significantly across groups (departments). Metrics As Scores computes this using Tukey's Honest Significance test.

Please have a look at the `About.pdf` that comes with the known datasets (the list is in the readme). For the prime example of the Qualitas.class corpus, we see that across groups (there: application domains) most software metrics are different (i.e., the groups distinguish clearly from each other). Please see <https://zenodo.org/record/7633950>.

---------------

**Edit**: I updated this comment and replaced *quantity type* and *context* with *feature* and *group*, respectively. I did implement this change in the application itself and also updated the known datasets on Zenodo to reflect that, as I believe that these should be the proper terms.

**Comment by mdhaber** [comment]
Thanks for the updates. 

I would characterize the examples in `README.md` as *when* (i.e. in what scenarios) to use the software, but the criterion requires "examples of *how* to use the software".  `README.md` describes what the raw data is like and questions you might ask, but to satisfy this criterion I'd expect a list of steps depicting the user's interaction with the software (e.g. including code and console output, screenshots, or video) interspersed with interpretations of what the user is doing and what the software's output means.

So for an "example" based on the diamonds dataset, using the web app, for example:

- What do you click to load the dataset? What will you see?
- "Therefore, we select only those diamonds from the dataset that have an ideal cut and the best (IF) clarity." *How* do you filter the dataset in the web app? 
- "Now only the color quality gives a context to each diamonds and its attributes (i.e., diamonds are now grouped by color)." *How do you group by color in the web app?
- "This constellation now allows us to examine differences across differently colored diamonds. For example, there are considerable differences in price." What do you look at in the web app to see this?

What would the steps be in the TUI?
What were the steps for producing a dataset to be analyzed (e.g. from a dataframe or CSV file)?

The same questions might be asked about other major features until the user has a sufficient understanding of the core concepts to generalize to other situations.

For instance, in the SciPy documentation examples (e.g. [`scipy.stats.jarque_bera`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.jarque_bera.html#scipy.stats.jarque_bera), [`scipy.stats.fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fit.html), [`scipy.stats.goodness_of_fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.goodness_of_fit.html), [`scipy.stats.tukey_hsd`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.tukey_hsd.html)), see how a description of what is being done is interspersed with code, code output, and an interpretation of that output.
[scikit learn](https://scikit-learn.org/stable/) and [matplotlib](https://matplotlib.org/stable/plot_types/index.html) also have great examples.

Since it would be a lot of work to do this for *all* functions, I'd suggest creating a separate "Examples" or "Tutorials" section of the documentation separate from the API documentation (but also separate from the Readme). There, you can give the user an overview of what is possible, and then the user will know enough to make sense of the documentation.

**Comment by MrShoenel** [comment]
I guess you are right, it's not *when*, but rather *how*. I will add some Jupyter notebooks to this repo that demonstrate the example usage.
I will comment here once it's done.

**Comment by MrShoenel** [comment]
There is now an additional new Jupyter notebook.
I have added a link to it in the README.md: <https://github.com/MrShoenel/metrics-as-scores#concrete-example-using-the-iris-dataset>
The example is quite detailed, so I hope it covers the usage quite well - the most important steps of the TUI are covered.

-----

**Edit**: The Iris dataset that is produced in this notebook is now also published and listed as a known dataset in the `README.md`.

**Comment by mdhaber** [comment]
Yes, I think it covers the TUI pretty well. If you were to write examples of all the other claimed uses, I think we could close this.

**Comment by MrShoenel** [comment]
By that, do you mean the remaining workflows of the TUI?
These are:
- Show Installed Datasets
- Show a List of Known Datasets Available Online
- Download and Install a Known Dataset
- Running the local Web Server

(show how all these can be done from code)

**Comment by mdhaber** [comment]
Not really. I was thinking of the web app. https://github.com/MrShoenel/metrics-as-scores/issues/5#issuecomment-1437450412

**Comment by MrShoenel** [comment]
I did add a new notebook, <https://github.com/MrShoenel/metrics-as-scores/blob/master/notebooks/Example-webapp-qcc.ipynb>.
It is also referenced in the README.
I do think that it illustrates the core concept quite well (getting scores).

I think it covers the claimed uses quite well now, what do you think?

**Comment by mdhaber** [comment]
1. I notice that at least one thing could use updating (e.g. the dataset can be selected without typing the ID), but I won't track it here.

2. In Part 4.2, my plot looks like:
![image](https://user-images.githubusercontent.com/6570539/224506208-542cf723-dc48-4e8e-8bf7-57cf6897f55a.png)

so I'm not seeing all the data that is shown in your plot.

3. Also, in 4.4 
![image](https://user-images.githubusercontent.com/6570539/224506243-41b7250c-e1f4-453f-8344-63ddc5ed86bf.png)

I don't see this option.

4. Consider adding a way to hide the table at the top or maybe move it to the bottom. At 100% zoom, I can't see the options at the top and the plot simultaneously, so throughout the tutorial I was scrolling back and forth between the two. 

5. IIRC, "Example Usage: Create Own Dataset" used to show how to use the TUI to prepare a dataset. Did you remove that part of the example? Now, it only shows how to use code.

6. Is the [documentation](https://mrshoenel.github.io/metrics-as-scores/) linked from the readme or anywhere else? Are the tutorials linked from the documentation?

**Comment by MrShoenel** [comment]
Hey,

> 1. I notice that at least one thing could use updating (e.g. the dataset can be selected without typing the ID), but I won't track it here.

That was fixed in 5825c13, or did you mean something else?


> 2. In Part 4.2, my plot looks like [...]:
> so I'm not seeing all the data that is shown in your plot.

Fair point, I have made some subtle changes to the example usage. Most notably, the box *Cut off smoothed distributions beyond observed values* shall not be ticked any longer. The image I had previously used in 4.2 had it unticked.


> 3. Also, in 4.4
>    ![image](https://user-images.githubusercontent.com/6570539/224506243-41b7250c-e1f4-453f-8344-63ddc5ed86bf.png)
> 
> I don't see this option.

Yes, that was confusing, too. It's the same input box as used in a previous step and I was referring to its placeholder text. That is now clear in the notebook!

 
> 4. Consider adding a way to hide the table at the top or maybe move it to the bottom. At 100% zoom, I can't see the options at the top and the plot simultaneously, so throughout the tutorial I was scrolling back and forth between the two.

I Agree, in 874d0a7 a toggle button for hiding the table was added and the controls were slightly re-arranged to be a little more space efficient.


> 5. IIRC, "Example Usage: Create Own Dataset" used to show how to use the TUI to prepare a dataset. Did you remove that part of the example? Now, it only shows how to use code.

That was not part of the notebook for "Example Usage: Create Own Dataset". However, I have added it now!

> 6. Is the [documentation](https://mrshoenel.github.io/metrics-as-scores/) linked from the readme or anywhere else? Are the tutorials linked from the documentation?

It is now; I have found a way to include the full README.md into the start page of the documentation, so now all is fully cross-linked!

**Comment by mdhaber** [comment]
> It is now; I have found a way to include the full README.md into the start page of the documentation, so now all is fully cross-linked!

You included the README.md in the documentation. I don't think it goes the other way.
Pretend that you are a user without a link to the documentation. How would you find the documentation?

> That was fixed in https://github.com/MrShoenel/metrics-as-scores/commit/5825c13145d20b6cad20ec9a19f6b2cda3545784, or did you mean something else?

I mean that the tutorial did not reflect that change. Now it does.

**Comment by MrShoenel** [comment]
> You included the README.md in the documentation. I don't think it goes the other way. Pretend that you are a user without a link to the documentation. How would you find the documentation?

It does because the README.md is at the repository's root, which has the link (see top right corner). But I have added an extra link within the readme, just to be certain (f129c51).

**Comment by mdhaber** [comment]
Note that when selecting a CSV file to transform into the required format, the TUI didn't like quotation marks around the path. I'd suggest that the TUI should accept paths surrounded by quotes (common for paths that include spaces).

**Comment by MrShoenel** [comment]
In 7a514f8, any leading and trailing single or double quotation marks are removed from the path. The quotation marks are not required, even for paths that contain spaces, because this is not an argument that is passed but rather a literal string input. In any case, even when passing a path enclosed in quotation marks, no error is thrown any longer.

**Comment by mdhaber** [comment]
Yes, I understand it was not required, but on Windows, "Copy as Path" includes quotation marks by default.

I'm trying to process this CSV using the TUI:
```
ID,Feature 1,Feature 2
1,0.130610668,-1.123509105
2,0.713853564,0.564677894
3,0.776684323,0.761043052
4,0.772784127,0.748047104
5,0.044091715,-1.705058948
6,0.567637633,0.170362917
7,0.020130102,-2.051069245
8,0.952960263,1.674260191
9,0.583233169,0.210171705
10,0.691309786,0.499566392
11,0.554604493,0.137303362
12,0.213504707,-0.794319569
13,0.245920098,-0.687384923
14,0.437125164,-0.158262029
15,0.936711533,1.527740684
16,0.094952442,-1.310860542
17,0.724460559,0.596144226
18,0.202796858,-0.831672697
19,0.726507354,0.602283726
20,0.281871831,-0.57728986
21,0.601629059,0.257565988
22,0.843001763,1.006871614
```
I select ID as the groups column and include both features.
I skip all the optional group descriptions. (It would be nice if they could all be skipped in one shot rather than pressing enter one time per group. If I hold it, it is easy to accidentally spilling over into the next prompt.) 

When trying to fit parametetric distributions, sometimes distribution names are repeated, and `_gen` appears after the name of each. I know why these things happen since I am a SciPy maintainer, but others may find it strange.

To speed things up, I deselected all distributions except for the normal distribution, and I get an error:
```
? Select continuous random variables: [norm_gen]
 -> Having selected 1 continuous random variables.
You can now select the discrete random variables that you want to
additionally attempt to fit to discrete features. Select all in
case you intend to redistribute and publicize your dataset.
While there are much fewer discrete random variables, their fitting
is dramatically more computationally expensive to compute, since a
global search has to be performed.
? Select discrete random variables: done
 -> Having selected 0 discrete random variables.
----------
You need to choose how fits should be computed. Metrics As Scores offers
two classes: Fitter and FitterPymoo, where the latter is recommended. As
far as continuous random variables are concerned, it does not make a
difference. For discrete random variables, however, the original Fitter
uses scipy's approach of differential evolution, which does not always
scale well. The class FitterPymoo, on the other hand, uses separate, mixed
variable genetic algorithm problems, tailored to each random variable.
? Which fitter do you want to use? FitterPymoo [Recommended]
----------
We are ready now to fit random variables. Between each of the transform types
(expectation/mean, median, mode, supremum, infimum) you get the chance to pause
and resume later, and the process will skip existing computed transforms. It is
recommended to run the following on a resourceful computer and to exploit the
highest possible degree of parallelism available. Next, you will be asked how
many cores you would like to use.
? How many cores should I use (1-16)? 2
----------
 -> Fits already computed for transform: <none> [NONE]
----------
 -> Generating parametric fits for: E[X] (expectation) [EXPECTATION]
 -> Performing distribution transforms for: continuous
100%|████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  6.31it/s]
An error (ValueError) has occurred: `dataset` input should have multiple elements.

Restarting main menu.
```

This happens whether I fit only the normal distribution or all the default selections.

**Comment by MrShoenel** [comment]
Good catch, because this is actually an invalid dataset as far as Metrics As Scores is concerned. A valid dataset needs to have at least two unique observations per feature and group, which is not the case here. The error originates in `KDE_approx` that uses `gaussian_kde()`, which does not work with a single observation. `KDE_approx`, however, is a class that is required for the web application.

I have made numerous commits and added testing to prevent users from using such datasets from now on (72316f1, 35751db, f2ef058, d9c1199, 90d7c2f, and 8c493f4). I used your dataset to generate a test case that should throw an informative exception.


------------

As for the things you raised:

> I skip all the optional group descriptions. (It would be nice if they could all be skipped in one shot rather than pressing enter one time per group. If I hold it, it is easy to accidentally spilling over into the next prompt.)

I get that, your example has uncommonly many groups and you are testing, hence the desire to skip things quickly. However, this workflow is intended for creating publishable datasets, and those fields really should be filled out :)

> When trying to fit parametetric distributions, sometimes distribution names are repeated, and _gen appears after the name of each. I know why these things happen since I am a SciPy maintainer, but others may find it strange.

~I do not get any repeated names, all of them end in `_gen` (I only use these objects). Which ones are repeated for you?~
Edit: Very few are repeated, but all and in `_gen`. Will investigate.
**Edit**: Only `trapezoid_gen` and `vonmises_gen` were repeated because these two have two (rather than one) configured instances (with different names). To fix the duplicate names, the list of continuous distributions now also shows the concrete instance's name, like `trapezoid_gen [trapz]` (a3f56a8).
